pkg/operator/controller/canary/controller.go

```diff
@@ -306,17 +311,48 @@ func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
 	return nil
 }
 
-func (r *reconciler) setCanaryFailingStatusCondition() error {
+func (r *reconciler) setCanaryFailingStatusCondition(errors []error) error {
+	errorStrings := DeduplicateErrorStrings(errors)
+	if len(errorStrings) > 3 {
+		errorStrings = errorStrings[len(errorStrings)-3:]
+	}
 	cond := operatorv1.OperatorCondition{
 		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
 		Status:  operatorv1.ConditionFalse,
 		Reason:  "CanaryChecksRepetitiveFailures",
-		Message: "Canary route checks for the default ingress controller are failing",
+		Message: fmt.Sprintf("Canary route checks for the default ingress controller are failing. Last %d error messages:\n%s", len(errorStrings), strings.Join(errorStrings, "\n")),
 	}
 
 	return r.setCanaryStatusCondition(cond)
 }
 
+// DeduplicateErrorStrings takes a slice of errors in chronological order, prunes the list for unique error messages
+// It returns a slice of the error message as strings, with any string seen more than once incuding how many times it
+// was seen
+// The chronological order is preserved, based on the last time each error message was seen
+func DeduplicateErrorStrings(errors []error) []string {
```

Comment:
Since it's now a public function and quite a generic one, it may deserve to be put into `pkg/util` package. That would justify the recently added unit test even more :)

Reply:
+1 to Andrey, but what's the motivation behind making this public now? Is it in anticipation for re-use? I suppose I assumed functions should be private unless they are used by other packages.

Reply:
I had assumed `pkg/operator/controller/canary/controller_test.go` was in the `test` package (which I believe is the idiomatic package to use in `*_test.go` files), which would require `DeduplicateErrorStrings` be public if it was to be tested. However, now that I look at it that's not the case; `controller_test.go` is in package `canary`.

I don't mind moving it to `pkg/util`, but since it's only used here at the moment, it would also be reasonable to just make it private again. @alebedev87 @gcs278 do you have a preference?

Reply:
I'm fine to keep it where it is, private and unit tested.

Reply:
+1 for just keeping private and unit tested.